Request to add additional kube_api_server metrics

expedia-hp opened the pull request:
### What does this PR do?
 It will add followings missing kube_api_server metrics

- apiserver_envelope_encryption_dek_cache_fill_percent
- aggregator_unavailable_apiservice
- apiserver_flowcontrol_request_concurrency_limit
- apiserver_flowcontrol_current_executing_requests
- apiserver_flowcontrol_rejected_requests_total 

### Motivation
<!-- What inspired you to submit this pull request? -->

### Additional Notes
<!-- Anything else we should know when reviewing? -->

### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests (unit, integration, e2e)
- [ ] PR title must be written as a CHANGELOG entry [(see why)](https://github.com/DataDog/integrations-core/blob/master/CONTRIBUTING.md#pull-request-title)
- [ ] Files changes must correspond to the primary purpose of the PR as described in the title (small unrelated changes should have their own PR)
- [ ] PR must have `changelog/` and `integration/` labels attached
- [ ] If the PR doesn't need to be tested during QA, please add a `qa/skip-qa` label.

clamoriniere commented on kube_apiserver_metrics/tests/test_kube_apiserver_metrics_1_25.py:
```diff
@@ -0,0 +1,115 @@
+
+@pytest.fixture()
+def mock_get():
+    f_name = os.path.join(os.path.dirname(__file__), 'fixtures', 'metrics_1.24.0.txt')
```
the file name say Kube version 1.25, but the metrics file is 1.24. Is it mandatory to use 1.25?

clamoriniere reviewed (commented):
thanks for opening this PR. 
It was in our roadmap to add apiservice metrics and flowcontrol. 

we need to review the state (beta, stable...) of this metrics to properly document it. and also check the cardinality.

iliakur commented on kube_apiserver_metrics/tests/test_kube_apiserver_metrics_1_25.py:
```diff
@@ -0,0 +1,115 @@
+        yield
+
+
+class TestKubeAPIServerMetrics:
```
Hmm, are you aware of #14147? This test looks exactly the same as the one in that PR... my guess is there will be merge conflicts, it may make sense to resolve them early.

expedia-hp commented:
@iliakur : What is the next step? Any update on this request, please.

clamoriniere commented on kube_apiserver_metrics/metadata.csv:
```diff
@@ -58,4 +58,10 @@ kube_apiserver.grpc_client_started_total,count,,request,,The total number of RPC
+kube_apiserver.apiserver_flowcontrol_current_executing_requests,gauge,,,,"Number of requests in initial (for a WATCH) or any (for a non-WATCH) execution stage in the API Priority and Fairness subsystem",0,kubernetes_api_server_metrics,flowcontrol request concurrency limit,
+kube_apiserver.apiserver_flowcontrol_request_concurrency_limit,gauge,,,,"Shared concurrency limit in the API Priority and Fairness subsystem",0,kubernetes_api_server_metrics, flowcontrol request concurrency limit,
+kube_apiserver.apiserver_flowcontrol_rejected_requests_total.count,count,,,,"Number of requests rejected by API Priority and Fairness subsystem",0,kubernetes_api_server_metrics, flowcontrol rejected requests total,
+kube_apiserver.kubernetes_feature_enabled,gauge,,,,"Whether a Kubernetes feature gate is enabled or not, identified by name and stage (alpha; Kubernetes 1.26+)",0,kubernetes_api_server_metrics,kubernetes feature enabled,
```
could you please remove this metric from the pull request because this metric was already added by https://github.com/DataDog/integrations-core/pull/14147 (and will be released in agent 7.45.0)

expedia-hp replied:
I have synced with the master branch and removed the one which were added recently. Test case is failing with different error. Can you please take a look and let me know?

clamoriniere reviewed (commented):
Hi @expedia-hp 

Sorry for the delay in the review. 
I added a comment to remove the one of the metrics that was added by another PR. 

also the unit-tests are failing, because the metrics that you have added was present in several kubernetes version already so you need to update the list of assertion metrics in 
test from 1.19 to 1.26. 

see: 
```
tests/test_kube_apiserver_metrics_1_19.py::TestKubeAPIServerMetrics::test_check FAILED [ 63%]
tests/test_kube_apiserver_metrics_1_23.py::TestKubeAPIServerMetrics::test_check FAILED [ 72%]
tests/test_kube_apiserver_metrics_1_24.py::TestKubeAPIServerMetrics::test_check FAILED [ 81%]
tests/test_kube_apiserver_metrics_1_25.py::TestKubeAPIServerMetrics::test_check FAILED [ 90%]
tests/test_kube_apiserver_metrics_1_26.py::TestKubeAPIServerMetrics::test_check FAILED [100%]
```

you can find the list of missing assertion metrics by searching `Found metrics that are not asserted` in the unittest logs here: https://github.com/DataDog/integrations-core/actions/runs/4798705021/jobs/8537361299?pr=14129

please ping me on the PR when the unit-tests are fixed or if you are not able to fix them. 🙇

expedia-hp commented:
Hi @Cedric ***@***.***>
I have added comments on PR. Please let me know If you have any questions.

Thanks
-Hiral

From: Cedric Lamoriniere ***@***.***>
Date: Thursday, April 27, 2023 at 4:41 AM
To: DataDog/integrations-core ***@***.***>
Cc: Hiral Patel ***@***.***>, Mention ***@***.***>
Subject: [External] Re: [DataDog/integrations-core] Request to add additional kube_api_server metrics (PR #14129)

@clamoriniere commented on this pull request.

Hi @expedia-hp<https://github.com/expedia-hp>

Sorry for the delay in the review.
I added a comment to remove the one of the metrics that was added by another PR.

also the unit-tests are failing, because the metrics that you have added was present in several kubernetes version already so you need to update the list of assertion metrics in
test from 1.19 to 1.26.

see:

tests/test_kube_apiserver_metrics_1_19.py::TestKubeAPIServerMetrics::test_check FAILED [ 63%]

tests/test_kube_apiserver_metrics_1_23.py::TestKubeAPIServerMetrics::test_check FAILED [ 72%]

tests/test_kube_apiserver_metrics_1_24.py::TestKubeAPIServerMetrics::test_check FAILED [ 81%]

tests/test_kube_apiserver_metrics_1_25.py::TestKubeAPIServerMetrics::test_check FAILED [ 90%]

tests/test_kube_apiserver_metrics_1_26.py::TestKubeAPIServerMetrics::test_check FAILED [100%]

you can find the list of missing assertion metrics by searching Found metrics that are not asserted in the unittest logs here: https://github.com/DataDog/integrations-core/actions/runs/4798705021/jobs/8537361299?pr=14129

please ping me on the PR when the unit-tests are fixed or if you are not able to fix them. 🙇

________________________________

In kube_apiserver_metrics/metadata.csv<https://github.com/DataDog/integrations-core/pull/14129#discussion_r1178804408>:

> @@ -58,4 +58,10 @@ kube_apiserver.grpc_client_started_total,count,,request,,The total number of RPC

 kube_apiserver.requested_deprecated_apis,gauge,,request,,"Gauge of deprecated APIs that have been requested, broken out by API group, version, resource, subresource, and removed_release",0,kubernetes_api_server_metrics,deprecated API requests,

 kube_apiserver.apiserver_admission_webhook_fail_open_count.count,count,,,,"Admission webhook fail open count, identified by name and broken out for each admission type (validating or mutating).",0,kubernetes_api_server_metrics,webhook fail open count,

 kube_apiserver.apiserver_admission_webhook_fail_open_count,gauge,,,,"Admission webhook fail open count, identified by name and broken out for each admission type (validating or mutating).",0,kubernetes_api_server_metrics,webhook fail open count,

+kube_apiserver.apiserver_envelope_encryption_dek_cache_fill_percent,gauge,,,,"Percent of the cache slots currently occupied by cached DEKs.",0,kubernetes_api_server_metrics,envelope encryption dek cache fillpercent,

+kube_apiserver.aggregator_unavailable_apiservice,gauge,,,,"Gauge of APIServices which are marked as unavailable broken down by APIService name.",0,kubernetes_api_server_metrics,aggregator unavailable apiservice,

+kube_apiserver.apiserver_flowcontrol_current_executing_requests,gauge,,,,"Number of requests in initial (for a WATCH) or any (for a non-WATCH) execution stage in the API Priority and Fairness subsystem",0,kubernetes_api_server_metrics,flowcontrol request concurrency limit,

+kube_apiserver.apiserver_flowcontrol_request_concurrency_limit,gauge,,,,"Shared concurrency limit in the API Priority and Fairness subsystem",0,kubernetes_api_server_metrics, flowcontrol request concurrency limit,

+kube_apiserver.apiserver_flowcontrol_rejected_requests_total.count,count,,,,"Number of requests rejected by API Priority and Fairness subsystem",0,kubernetes_api_server_metrics, flowcontrol rejected requests total,

+kube_apiserver.kubernetes_feature_enabled,gauge,,,,"Whether a Kubernetes feature gate is enabled or not, identified by name and stage (alpha; Kubernetes 1.26+)",0,kubernetes_api_server_metrics,kubernetes feature enabled,

could you please remove this metric from the pull request because this metric was already added by #14147<https://github.com/DataDog/integrations-core/pull/14147> (and will be released in agent 7.45.0)

—
Reply to this email directly, view it on GitHub<https://github.com/DataDog/integrations-core/pull/14129#pullrequestreview-1403473640>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/A6HT5G4WR2DHRNWAREZ35W3XDIWJZANCNFSM6AAAAAAVVQEPHE>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

clamoriniere commented:
Hi @expedia-hp,

The changes to fix the CI failures were not always trivial, so I decided to do a new PR based on yours, to fix the unit-test and apply naming convention on the new metric names.

You can review it https://github.com/DataDog/integrations-core/pull/14480 to see the changes were needed.

You can close after this PR.
thanks again for this contribution, Adding `flowcontrol` metrics was already under our radar, but not prioritised yet.

clamoriniere commented:
I think we can close this pull request since all the metrics that this PR introduce was added by other contributions.
